File: scm-csharp/scheme/SchemeHttpResponse.cs
using System.Collections.Generic;
using System.Text;

namespace scheme;

public class SchemeHttpResponse
{
    public int Status;
    public List<(string Name, string Value)> Headers;
    public string Body;
    public byte[]? BodyBytes;

    public SchemeHttpResponse(int status, List<(string, string)> headers, string body)
    {
        Status = status;
        Headers = headers;
        Body = body;
        BodyBytes = null;
    }

    public SchemeHttpResponse(int status, List<(string, string)> headers, byte[] bodyBytes)
    {
        Status = status;
        Headers = headers;
        Body = "";
        BodyBytes = bodyBytes;
    }

    public byte[] GetBodyBytes() => BodyBytes ?? Encoding.UTF8.GetBytes(Body);

    public override string ToString() => $"#<http-response {Status}>";
}

File: scm-java/src/main/scheme/SchemeHttpResponse.java
package scheme;

import java.nio.charset.StandardCharsets;
import java.util.List;

public class SchemeHttpResponse {
    public int status;
    public List<String[]> headers; // each entry: [name, value]
    public String body;
    public byte[] bodyBytes;

    public SchemeHttpResponse(int status, List<String[]> headers, String body) {
        this.status = status;
        this.headers = headers;
        this.body = body;
        this.bodyBytes = null;
    }

    public SchemeHttpResponse(int status, List<String[]> headers, byte[] bodyBytes) {
        this.status = status;
        this.headers = headers;
        this.body = "";
        this.bodyBytes = bodyBytes;
    }

    public byte[] getBodyBytes() {
        return bodyBytes != null ? bodyBytes : body.getBytes(StandardCharsets.UTF_8);
    }

    @Override
    public String toString() { return "#<http-response " + status + ">"; }
}
